File: phone/app/src/main/java/com/parado/smsbridge/protocol/Protocol.kt
package com.parado.smsbridge.protocol

import org.json.JSONObject

/**
 * 两端共享的局域网消息协议，须与 computer/src/protocol.rs 保持一致。
 *
 * 安全约束：消息体只携带必要字段，**绝不包含短信原文、手机号或发件人**。
 * 使用 Android 内置的 org.json，不引入任何外部依赖。
 */
object Protocol {

    const val VERSION = 1

    const val TYPE_PAIR_REQUEST = "pair_request"
    const val TYPE_PAIR_RESPONSE = "pair_response"
    const val TYPE_CODE = "code"
    const val TYPE_HEARTBEAT = "heartbeat"
    const val TYPE_UNPAIR = "unpair"
    const val TYPE_DISCOVERY_REQUEST = "discovery_request"
    const val TYPE_DISCOVERY_RESPONSE = "discovery_response"

    sealed class Message {
        abstract val v: Int

        /** 手机 → 电脑：携带配对码的配对请求 */
        data class PairRequest(
            override val v: Int,
            val deviceId: String,
            val code: String,
        ) : Message()

        /** 电脑 → 手机：配对结果 */
        data class PairResponse(
            override val v: Int,
            val ok: Boolean,
            val sessionId: String?,
        ) : Message()

        /** 手机 → 电脑：验证码（只含验证码本身） */
        data class Code(
            override val v: Int,
            val code: String,
            val ts: Long,
            val nonce: String,
            val mac: String,
        ) : Message()

        /** 手机 → 电脑：心跳保活（仅含版本与设备号） */
        data class Heartbeat(
            override val v: Int,
            val deviceId: String,
        ) : Message()

        /** 手机 → 电脑（广播）：谁在线？ */
        data class DiscoveryRequest(
            override val v: Int,
            val deviceId: String,
        ) : Message()

        /** 电脑 → 手机：我在这里（仅含发现所需字段） */
        data class DiscoveryResponse(
            override val v: Int,
            val deviceId: String,
            val name: String,
            val port: Int,
        ) : Message()

        /** 手机 → 电脑：主动解绑（清密钥 + 清配对） */
        data class Unpair(
            override val v: Int,
            val deviceId: String,
        ) : Message()
    }

    /** 版本不匹配的消息不得进入后续处理流程。 */
    fun isSupported(message: Message): Boolean = message.v == VERSION

    fun toJson(message: Message): String = JSONObject().apply {
        when (message) {
            is Message.PairRequest -> {
                put("type", TYPE_PAIR_REQUEST)
                put("v", message.v)
                put("device_id", message.deviceId)
                put("code", message.code)
            }

            is Message.PairResponse -> {
                put("type", TYPE_PAIR_RESPONSE)
                put("v", message.v)
                put("ok", message.ok)
                if (message.sessionId == null) {
                    put("session_id", JSONObject.NULL)
                } else {
                    put("session_id", message.sessionId)
                }
            }

            is Message.Code -> {
                put("type", TYPE_CODE)
                put("v", message.v)
                put("code", message.code)
                put("ts", message.ts)
                put("nonce", message.nonce)
                put("mac", message.mac)
            }

            is Message.Heartbeat -> {
                put("type", TYPE_HEARTBEAT)
                put("v", message.v)
                put("device_id", message.deviceId)
            }

            is Message.DiscoveryRequest -> {
                put("type", TYPE_DISCOVERY_REQUEST)
                put("v", message.v)
                put("device_id", message.deviceId)
            }

            is Message.DiscoveryResponse -> {
                put("type", TYPE_DISCOVERY_RESPONSE)
                put("v", message.v)
                put("device_id", message.deviceId)
                put("name", message.name)
                put("port", message.port)
            }

            is Message.Unpair -> {
                put("type", TYPE_UNPAIR)
                put("v", message.v)
                put("device_id", message.deviceId)
            }
        }
    }.toString()

    /** 类型未知或结构非法时返回 null，由调用方丢弃。 */
    fun parse(text: String): Message? = try {
        val json = JSONObject(text)
        val v = json.optInt("v", -1)
        when (json.optString("type")) {
            TYPE_PAIR_REQUEST -> Message.PairRequest(
                v,
                json.getString("device_id"),
                json.getString("code"),
            )

            TYPE_PAIR_RESPONSE -> Message.PairResponse(
                v,
                json.optBoolean("ok", false),
                if (json.isNull("session_id")) null else json.optString("session_id"),
            )

            TYPE_CODE -> Message.Code(
                v,
                json.getString("code"),
                json.getLong("ts"),
                json.getString("nonce"),
                json.getString("mac"),
            )

            TYPE_HEARTBEAT -> Message.Heartbeat(
                v,
                json.getString("device_id"),
            )

            TYPE_DISCOVERY_REQUEST -> Message.DiscoveryRequest(
                v,
                json.getString("device_id"),
            )

            TYPE_DISCOVERY_RESPONSE -> Message.DiscoveryResponse(
                v,
                json.getString("device_id"),
                json.optString("name"),
                json.optInt("port", 0),
            )

            TYPE_UNPAIR -> Message.Unpair(
                v,
                json.getString("device_id"),
            )

            else -> null
        }
    } catch (_: Exception) {
        null
    }
}

File: computer/src/protocol.rs
//! 两端共享的局域网消息协议。
//!
//! 安全约束：消息体只携带必要字段，**绝不包含短信原文、手机号或发件人**。
//! 夹具见 `shared/testdata/`，与 Android 端共用，保证两端解析行为一致。

use serde::{Deserialize, Serialize};

/// 协议版本。版本不一致的消息一律拒绝处理。
pub const PROTOCOL_VERSION: u32 = 1;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Message {
    /// 手机 → 电脑：携带配对码的配对请求
    PairRequest {
        v: u32,
        device_id: String,
        code: String,
        /// 手机生成的随机盐（hex）。两端各自用 code + salt 派生同一会话密钥，密钥本身不上网传输。
        salt_hex: String,
    },
    /// 电脑 → 手机：配对结果
    PairResponse {
        v: u32,
        ok: bool,
        session_id: Option<String>,
    },
    /// 手机 → 电脑：验证码（只含验证码本身）
    Code {
        v: u32,
        code: String,
        ts: i64,
        nonce: String,
        mac: String,
    },
    /// 手机 → 电脑：心跳保活（仅含版本与设备号，绝不携带敏感信息）
    Heartbeat { v: u32, device_id: String },
    /// 手机 → 电脑（广播）：谁在线？
    DiscoveryRequest { v: u32, device_id: String },
    /// 电脑 → 手机：我在这里（仅含发现所需字段）
    DiscoveryResponse {
        v: u32,
        device_id: String,
        name: String,
        port: u16,
    },
    /// 手机 → 电脑：主动解绑（清密钥 + 清配对）
    Unpair { v: u32, device_id: String },
}

impl Message {
    pub fn version(&self) -> u32 {
        match self {
            Message::PairRequest { v, .. }
            | Message::PairResponse { v, .. }
            | Message::Code { v, .. }
            | Message::Heartbeat { v, .. }
            | Message::DiscoveryRequest { v, .. }
            | Message::DiscoveryResponse { v, .. }
            | Message::Unpair { v, .. } => *v,
        }
    }

    /// 版本不匹配的消息不得进入后续处理流程。
    pub fn is_supported(&self) -> bool {
        self.version() == PROTOCOL_VERSION
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }

    pub fn from_json(text: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(text)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pair_request_roundtrip() {
        let raw = include_str!("../../shared/testdata/pair_request.json");
        let message = Message::from_json(raw).expect("解析配对请求失败");
        match &message {
            Message::PairRequest {
                code, device_id, salt_hex, ..
            } => {
                assert_eq!(code, "123456");
                assert_eq!(device_id, "device-0001");
                assert_eq!(salt_hex, "30313233343536373839616263646566");
            }
            other => panic!("消息类型不符: {other:?}"),
        }
        assert!(message.is_supported());
        let encoded = message.to_json().expect("序列化失败");
        assert_eq!(Message::from_json(&encoded).expect("回读失败"), message);
    }

    #[test]
    fn pair_response_roundtrip() {
        let raw = include_str!("../../shared/testdata/pair_response.json");
        let message = Message::from_json(raw).expect("解析配对响应失败");
        match &message {
            Message::PairResponse { ok, session_id, .. } => {
                assert!(ok);
                assert_eq!(session_id.as_deref(), Some("s-0001"));
            }
            other => panic!("消息类型不符: {other:?}"),
        }
    }

    #[test]
    fn code_message_roundtrip() {
        let raw = include_str!("../../shared/testdata/code_message.json");
        let message = Message::from_json(raw).expect("解析验证码消息失败");
        match &message {
            Message::Code {
                code, ts, nonce, ..
            } => {
                assert_eq!(code, "482913");
                assert_eq!(*ts, 1_757_337_600_000);
                assert_eq!(nonce, "n-0001");
            }
            other => panic!("消息类型不符: {other:?}"),
        }
        let encoded = message.to_json().expect("序列化失败");
        assert_eq!(Message::from_json(&encoded).expect("回读失败"), message);
    }

    #[test]
    fn discovery_request_roundtrip() {
        let raw = include_str!("../../shared/testdata/discovery_request.json");
        let message = Message::from_json(raw).expect("解析发现请求失败");
        match &message {
            Message::DiscoveryRequest { device_id, .. } => assert_eq!(device_id, "device-0001"),
            other => panic!("消息类型不符: {other:?}"),
        }
        let encoded = message.to_json().expect("序列化失败");
        assert_eq!(Message::from_json(&encoded).expect("回读失败"), message);
    }

    #[test]
    fn discovery_response_roundtrip() {
        let raw = include_str!("../../shared/testdata/discovery_response.json");
        let message = Message::from_json(raw).expect("解析发现响应失败");
        match &message {
            Message::DiscoveryResponse {
                device_id,
                name,
                port,
                ..
            } => {
                assert_eq!(device_id, "pc-0001");
                assert_eq!(name, "PC-0001");
                assert_eq!(*port, 45876);
            }
            other => panic!("消息类型不符: {other:?}"),
        }
        let encoded = message.to_json().expect("序列化失败");
        assert_eq!(Message::from_json(&encoded).expect("回读失败"), message);
    }

    #[test]
    fn unknown_field_is_ignored() {
        let raw = include_str!("../../shared/testdata/code_message_unknown_field.json");
        let message = Message::from_json(raw).expect("未知字段应当被忽略");
        let expected = Message::from_json(include_str!("../../shared/testdata/code_message.json"))
            .expect("解析基准夹具失败");
        assert_eq!(message, expected);
    }

    #[test]
    fn version_mismatch_rejected() {
        let raw = include_str!("../../shared/testdata/code_message_version_mismatch.json");
        let message = Message::from_json(raw).expect("结构合法，应能解析");
        assert!(!message.is_supported());
    }

    #[test]
    fn code_message_never_carries_sensitive_fields() {
        let raw = include_str!("../../shared/testdata/code_message.json");
        let message = Message::from_json(raw).expect("解析失败");
        let encoded = message.to_json().expect("序列化失败");
        for forbidden in ["body", "sender", "phone", "address", "content"] {
            assert!(
                !encoded.contains(forbidden),
                "消息体不应出现字段 {forbidden}: {encoded}"
            );
        }
    }

    #[test]
    fn heartbeat_roundtrip() {
        let raw = include_str!("../../shared/testdata/heartbeat.json");
        let message = Message::from_json(raw).expect("解析心跳失败");
        match &message {
            Message::Heartbeat { v, device_id } => {
                assert_eq!(*v, PROTOCOL_VERSION);
                assert_eq!(device_id, "device-0001");
            }
            other => panic!("消息类型不符: {other:?}"),
        }
        assert!(message.is_supported());
        let encoded = message.to_json().expect("序列化失败");
        assert_eq!(Message::from_json(&encoded).expect("回读失败"), message);
    }

    #[test]
    fn unpair_roundtrip() {
        let raw = include_str!("../../shared/testdata/unpair.json");
        let message = Message::from_json(raw).expect("解析解绑消息失败");
        match &message {
            Message::Unpair { v, device_id } => {
                assert_eq!(*v, PROTOCOL_VERSION);
                assert_eq!(device_id, "device-0001");
            }
            other => panic!("消息类型不符: {other:?}"),
        }
        let encoded = message.to_json().expect("序列化失败");
        assert_eq!(Message::from_json(&encoded).expect("回读失败"), message);
    }

    #[test]
    fn heartbeat_and_unpair_never_carry_sensitive_fields() {
        let heartbeat = Message::Heartbeat {
            v: PROTOCOL_VERSION,
            device_id: "device-0001".into(),
        };
        let unpair = Message::Unpair {
            v: PROTOCOL_VERSION,
            device_id: "device-0001".into(),
        };
        let request = Message::DiscoveryRequest {
            v: PROTOCOL_VERSION,
            device_id: "device-0001".into(),
        };
        let response = Message::DiscoveryResponse {
            v: PROTOCOL_VERSION,
            device_id: "pc-0001".into(),
            name: "PC-0001".into(),
            port: 45876,
        };
        for (name, message) in [
            ("heartbeat", heartbeat),
            ("unpair", unpair),
            ("discovery_request", request),
            ("discovery_response", response),
        ] {
            let encoded = message.to_json().expect("序列化失败");
            for forbidden in [
                "body", "sender", "phone", "address", "content", "secret", "code",
            ] {
                assert!(
                    !encoded.contains(forbidden),
                    "{name} 不应出现字段 {forbidden}: {encoded}"
                );
            }
        }
    }
}
